oauth2_http/java/com/google/auth/oauth2/AwsRequestSigner.java

```diff
@@ -0,0 +1,362 @@
+    }
+    canonicalRequest.append(canonicalHeaders).append("\n");
+
+    // Append the signed headers.
+    canonicalRequest.append(Joiner.on(';').join(sortedHeaderNames)).append("\n");
+
+    // Append the hashed request payload.
+    canonicalRequest.append(getHexEncodedSha256Hash(requestPayload.getBytes()));
+
+    // Return the hashed canonical request.
+    return getHexEncodedSha256Hash(canonicalRequest.toString().getBytes());
+  }
+
+  /** Task 2: Create a string to sign for Signature Version 4. */
+  private String createStringToSign(
+      String canonicalRequestHash, String xAmzDate, String credentialScope) {
+    return HASHING_ALGORITHM
+        + "\n"
+        + xAmzDate
+        + "\n"
+        + credentialScope
+        + "\n"
+        + canonicalRequestHash;
+  }
+
+  /**
+   * Task 3: Calculate the signature for AWS Signature Version 4.
+   *
+   * @param date the date used in the hashing process in YYYYMMDD format
+   */
+  private String calculateAwsV4Signature(
+      String serviceName, String secret, String date, String region, String stringToSign) {
+    byte[] kDate = sign(("AWS4" + secret).getBytes(UTF_8), date.getBytes(UTF_8));
+    byte[] kRegion = sign(kDate, region.getBytes(UTF_8));
+    byte[] kService = sign(kRegion, serviceName.getBytes(UTF_8));
+    byte[] kSigning = sign(kService, AWS_REQUEST_TYPE.getBytes(UTF_8));
+    return BaseEncoding.base16().lowerCase().encode(sign(kSigning, stringToSign.getBytes(UTF_8)));
+  }
+
+  private Map<String, String> getCanonicalHeaders(String date) {
+    Map<String, String> headers = new HashMap<>();
+    headers.put("host", uri.getHost());
+
+    // Only add the date if it hasn't been specified through the "date" header.
+    if (!additionalHeaders.containsKey("date")) {
+      headers.put("x-amz-date", date);
+    }
+
+    if (awsSecurityCredentials.getToken() != null && !awsSecurityCredentials.getToken().isEmpty()) {
+      headers.put("x-amz-security-token", awsSecurityCredentials.getToken());
+    }
+
+    // Add all additional headers.
+    for (String key : additionalHeaders.keySet()) {
+      // Header keys need to be lowercase.
+      headers.put(key.toLowerCase(), additionalHeaders.get(key));
+    }
+    return headers;
+  }
+
+  private AwsDates getDates() {
+    if (additionalHeaders.containsKey("date")) {
+      return AwsDates.fromDateHeader(additionalHeaders.get("date"));
+    }
+    if (additionalHeaders.containsKey("x-amz-date")) {
+      return AwsDates.fromXAmzDate(additionalHeaders.get("x-amz-date"));
+    }
+    return AwsDates.generateXAmzDate();
+  }
+
+  private static byte[] sign(byte[] key, byte[] value) {
+    try {
+      String algorithm = "HmacSHA256";
+      Mac mac = Mac.getInstance(algorithm);
+      mac.init(new SecretKeySpec(key, algorithm));
+      return mac.doFinal(value);
+    } catch (NoSuchAlgorithmException | InvalidKeyException ex) {
+      throw new IllegalStateException("Failed to calculate the AWS V4 Signature.", ex);
```

Comment:
Pretty sure this is not the right exception here. 

Also, what is the justification for converting this to a runtime exception?

Reply:
What changes would you like to see here?

Reply:
Depends on whether a runtime exception is called for here or not. What does it mean if HmacSHA256 is not supported? What does it mean if the key is invalid? Where does the key come from?

Reply:
https://docs.oracle.com/javase/7/docs/api/javax/crypto/Mac.html

"Every implementation of the Java platform is required to support the following standard Mac algorithms:

HmacMD5
HmacSHA1
HmacSHA256"

I'll use a RuntimeException here.